slam-libraries/viam-cartographer/src/slam_service/slam_service.h

```diff
@@ -162,24 +145,88 @@ class SLAMServiceImpl final : public SLAMService::Service {
     // timestamp of the time when the map is saved.
     void SaveMapWithTimestamp();
 
-    SLAMServiceActionMode action_mode = SLAMServiceActionMode::MAPPING;
+    // GetJpegMap paints the jpeg version of the map and writes the
+    // image to the response. Returns a grpc status that reflects
+    // whether or not painting the map and writing it to the response
+    // was successful.
+    ::grpc::Status GetJpegMap(const GetMapRequest *request,
+                              GetMapResponse *response);
+
+    // GetPointCloudMap writes the pointcloud version of the map to
+    // the response. Returns a grpc status that reflects
+    // whether or not writing the map to the response
+    // was successful.
+    ::grpc::Status GetPointCloudMap(const GetMapRequest *request,
+                                    GetMapResponse *response);
+
+    // ProcessDataAndStartSavingMaps processes the data in the data directory
+    // that is newer than the provided data_cutoff_time
+    // and starts the process to save maps in parallel. In offline mode,
+    // all data in the directory is processed. In online mode, the most
+    // recently generated data is processed until a shutdown signal is
+    // received.
+    void ProcessDataAndStartSavingMaps(double data_cutoff_time);
+
+    // SetUpMapBuilder loads the lua file with default cartographer config
+    // parameters depending on the action mode.
+    void SetUpMapBuilder();
+
+    // GetLatestJpegMapString paints and returns the latest map as a jpeg string
+    // with or without the pose marker depending on the argument value.
+    std::string GetLatestJpegMapString(bool add_pose_marker);
```

Comment:
Please add comments to functions that are expected to be called with a particular mutex held.

Reply:
I thought it is required that `optimization_shared_mutex` is held while calling this function, since otherwise this function is unacceptably slow.

Reply:
Ah! The function is slow but can be called irrespectively `optimization_shared_mutex`. The `optimization_shared_mutex` checks merely when the optimization is running - and if it is running, then we're returning the maps we generated and saved before the optimization started.